src/runtime/GlobalObjectBuiltinPromise.cpp

```diff
@@ -357,6 +357,42 @@ static Value builtinPromiseCatch(ExecutionState& state, Value thisValue, size_t
 }
 
 
+static Value builtinPromiseFinally(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
+{
+    auto strings = &state.context()->staticStrings();
+    if (!thisValue.isObject())
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, strings->Promise.string(), false, strings->then.string(), "%s: not a Promise object");
+    Object* thisObject = thisValue.toObject(state);
+    Value onFinally = argv[0];
+    FunctionObject* thenFinally;
+    FunctionObject* catchFinally;
+    Value arguments[2];
+    Value C = thisObject->speciesConstructor(state, state.context()->globalObject()->promise());
+    if (onFinally.isPointerValue() && onFinally.asPointerValue()->isCallable()) {
+        thenFinally = new NativeFunctionObject(state, NativeFunctionInfo(strings->Empty, promiseThenFinally, 1, NativeFunctionInfo::Strict));
+        thenFinally->deleteOwnProperty(state,strings->name);
+        Object* thenInternalSlot = new Object(state);
+        thenFinally->setInternalSlot(thenInternalSlot);
```

Comment:
I'd prefer storing this data in the NativeFunction object. However it does not have member for this purpose, so I'd create a `PromiseStepFunctionObject` class inherited from the `NativeFunctionObject`. Via this solution no extra object allocation is needed.